ENIGMAsystem/SHELL/Graphics_Systems/General/GSmodel.cpp

```diff
@@ -195,9 +198,18 @@ void d3d_model_draw(int id) {
 }
 
 void d3d_model_draw(int id, gs_scalar x, gs_scalar y, gs_scalar z) {
-  d3d_transform_add_translation(x, y, z);
+  // backup the current world matrix
+  d3d_transform_stack_push();
+
+  // we have to create a special translation here so that it occurs
+  // before any of the user's transformations took place
+  enigma::world = glm::translate(enigma::world, glm::vec3(x, y, z));
+  enigma::graphics_set_matrix(matrix_world);
+
   d3d_model_draw(id);
-  d3d_transform_add_translation(-x, -y, -z);
+
+  // restore the world matrix the user had before this call
+  d3d_transform_stack_pop();
```

Comment:
This does not reset `matrix_world`. Future operations will still apply to the CPU-side matrix. Are you sure that's not what's causing this problem in the first place?

Reply:
You have two matrix representations. You have a GLM-managed matrix in RAM, and you have an actual matrix in VRAM used by DX. You are only pushing and popping the GPU matrix. It works because you're clearing the matrix before applying transformations, which, if the rest of the code looks like this, you are never doing anywhere else...

Reply:
> You are only pushing and popping the GPU matrix.

Incorrect, I am pushing and popping the CPU matrix. We implement the transform stack on the CPU, not the GPU. We could change that later, I'll do an apitrace real fast to see where GM does the transform stack.
https://github.com/enigma-dev/enigma-dev/blob/b39e88c2375e14b7dfdc97e2b12044303c8e4d7d/ENIGMAsystem/SHELL/Graphics_Systems/General/GSmatrix.cpp#L371
https://github.com/enigma-dev/enigma-dev/blob/b39e88c2375e14b7dfdc97e2b12044303c8e4d7d/ENIGMAsystem/SHELL/Graphics_Systems/General/GSmatrix.cpp#L387

Reply:
Oh, this is your implementation of that routine. Okay.